My NUnit

bulat-tsydendorzhiev opened the pull request.

bulat-tsydendorzhiev committed:
feat: add attributes

bulat-tsydendorzhiev committed:
feat: add test runner code

bulat-tsydendorzhiev committed:
feat: add prettier structure of folders, printer for report of tests
test: add simple tests;
fix: add assignment for ErrorMessage and IgnoreMessage

bulat-tsydendorzhiev committed:
fix: fix problem with exceptions, change folder style
feat: add new status

bulat-tsydendorzhiev committed:
feat: add check for valid before and after class attributes methods;
test: add invalid test cases

bulat-tsydendorzhiev committed:
feat: add interaction with user;
fix: change error message for before and after attributes exception occurrence

yurii-litvinov commented on MyNUnit/MyNUnit.Tests/MyNUnit.Tests.csproj:
```diff
@@ -0,0 +1,23 @@
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
```
Сюда тоже стоило подключить StyleCop

yurii-litvinov commented on MyNUnit/MyNUnit/MyNUnit.csproj:
```diff
@@ -0,0 +1,21 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
```
Используйте 1.2-beta, 1.1 не знает современного C#

yurii-litvinov commented on MyNUnit/MyNUnit/MyNUnit/TestRunner.cs:
```diff
@@ -0,0 +1,36 @@
+
+        var result = new List<TestAssemblyResult>();
+
+        Parallel.ForEach(assemblies, assembly => result.Add(assembly.RunTests()));
```
result не потокобезопасный, так что вызывать result.Add без синхронизации нельзя — внутри самого списка могут быть гонки, которые испортят данные

yurii-litvinov commented on MyNUnit/MyNUnit/MyNUnit/TestRunner.cs:
```diff
@@ -0,0 +1,36 @@
+
+        Parallel.ForEach(assemblies, assembly => result.Add(assembly.RunTests()));
+
+        return result.ToList();
```
Он же и так List?

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestClass.cs:
```diff
@@ -0,0 +1,93 @@
+        _testMethods = GetMethodsWithAttribute(typeof(MyTestAttribute), methods)
+            .Select(method =>
+            {
+                var instance = Activator.CreateInstance(testingClass);
```
Activator.CreateInstance неправильно вызывать до того, как _beforeClassMethods отработают. В них может быть какая-то настройка окружения, необходимого для корректной работы конструктора

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestClass.cs:
```diff
@@ -0,0 +1,93 @@
+            {
+                var instance = Activator.CreateInstance(testingClass);
+
+                ArgumentNullException.ThrowIfNull(instance);
```
Это же не аргумент, тут какое-то другое исключение надо бросать

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestClass.cs:
```diff
@@ -0,0 +1,93 @@
+                _invalidMethods.Add(method.Name);
+            }
+        }
+    }
```
Раз уж что-то проверяете, стоит проверять и что у методов нет параметров и тип возвращаемого значения — void, и для обычных тестовых методов тоже проверять всё это (и что они не static)

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestComponentsResults/TestAssemblyResult.cs:
```diff
@@ -0,0 +1,39 @@
+/// <summary>
+/// Defines the result of the classes test methods.
+/// </summary>
+public record TestAssemblyResult
```
Если это record, почему бы не использовать primary constructor и не писать свойства вовсе? Для record-ов их компилятор сгенерирует по параметрам основного конструктора.

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestMethod.cs:
```diff
@@ -0,0 +1,113 @@
+            status = errorMessage == string.Empty
+                ? MyTestStatus.Passed
+                : MyTestStatus.Failed;
+        }
```
А если исключение ожидалось, но не было брошено?

bulat-tsydendorzhiev committed:
feat: add check for case with expected exception but without throwing, add new stylecop version;
fix: add invokes for before class methods before creating the activator, change result on thread-safe option;
refactor: change representation of results of test components

bulat-tsydendorzhiev committed:
feat: add check for correctness of methods, add new canceled case,printer can only print assemblies with tests;
test: add new invalid tests

bulat-tsydendorzhiev committed:
fix: make correct name in tests

bulat-tsydendorzhiev committed:
style: decided to leave "cancelled"

yurii-litvinov commented on MyNUnit/MyNUnit/Attributes/AfterAttribute.cs:
```diff
@@ -0,0 +1,19 @@
+/// Identifies a method to be called immediately after each test is run.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+public class AfterAttribute : Attribute
```
Атрибуты и MyAssert лучше вынести в отдельную сборку, чтобы при написании тестов не надо было зависеть от всего MyNUnit

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestClass.cs:
```diff
@@ -0,0 +1,168 @@
+
+        _testMethods = testMethods.Select(method =>
+                                        {
+                                            var instance = Activator.CreateInstance(testingClass) ?? throw new NullReferenceException();
```
NullReferenceException сами не бросайте никогда, оно зарезервировано для рантайма .NET. Тут можно было своё какое-то исключение бросать

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestComponentsResults/TestAssemblyResult.cs:
```diff
@@ -0,0 +1,15 @@
+/// <param name="Name">Gets the test name of assembly with the classes with the tests.</param>
+/// <param name="ClassesTestResults">Gets the results of all tests of all classes in this assembly.</param>
+/// <param name="TotalDuration">Gets the time spent on all tests of all classes in this assembly in milliseconds.</param>
+public record TestAssemblyResult(string Name, IEnumerable<TestClassResult> ClassesTestResults, long TotalDuration)
+{
+}
```
```suggestion
public record TestAssemblyResult(string Name, IEnumerable<TestClassResult> ClassesTestResults, long TotalDuration);
```

yurii-litvinov commented on MyNUnit/MyNUnit/TestComponents/TestMethod.cs:
```diff
@@ -0,0 +1,124 @@
+
+        if (_expectedExceptionType is not null && errorMessage is null)
+        {
+            errorMessage = $"Expected {_expectedExceptionType} but it didn't occur";
```
Немного костыльно, можно было прямо в try эту проверку перенести. Но ок, так тоже можно

yurii-litvinov reviewed (approved):
В целом теперь всё ок, зачтена.
